netaddr.go

```diff
@@ -1263,31 +1288,43 @@ func (p IPPrefix) Overlaps(o IPPrefix) bool {
 	return p.ip == o.ip
 }
 
-// MarshalText implements the encoding.TextMarshaler interface,
-// The encoding is the same as returned by String, with one exception:
-// If p is the zero value, the encoding is the empty string.
-func (p IPPrefix) MarshalText() ([]byte, error) {
+// AppendTo appends a text encoding of p,
+// as generated by MarshalText,
+// to b and returns the extended buffer.
+func (p IPPrefix) AppendTo(b []byte) []byte {
 	if p.IsZero() {
-		return []byte(""), nil
+		return b
 	}
 	if !p.Valid() {
-		return []byte("invalid IP prefix"), nil
+		return append(b, "invalid IP prefix"...)
 	}
 
 	// p.IP is non-zero, because p is valid.
-	var b []byte
 	if p.ip.z == z4 {
-		max := len("255.255.255.255/32")
-		b = make([]byte, 0, max)
 		b = p.ip.appendTo4(b)
 	} else {
-		max := len("ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff%enp5s0/128")
-		b = make([]byte, 0, max)
 		b = p.ip.appendTo6(b)
 	}
 
 	b = append(b, '/')
 	b = appendDecimal(b, p.bits)
+	return b
+}
+
+// MarshalText implements the encoding.TextMarshaler interface,
+// The encoding is the same as returned by String, with one exception:
+// If p is the zero value, the encoding is the empty string.
+func (p IPPrefix) MarshalText() ([]byte, error) {
+	var max int
+	switch p.ip.z {
+	case z0:
```

Comment:
Could you return nil here, or does MarshalText mandate returning a non-nil zero length slice?

Reply:
Good question. I don't see anything in the docs that require non-nil. I'm going to leave as-is here to avoid changing behavior at the same time as adding API. I'll send a follow-up shortly.

Reply:
This turned out to be a really good question. Our marshaling behavior is rather inconsistent and probably incorrect in some cases.